Bot tweaks

Superlagg opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- To automatically tag this PR, add the uppercase label(s) surrounded by brackets below, for example: [LABEL] -->
[FEATURE] [REWORK]
## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

Made some tweaks to bots.

Across all bots:
- Bot processing tier is variable, using spatial hashing to determine if anyone's nearby.
- - If nobody's near, the bot's processing tier drops to an idle state.
- - If someone's near, the bot bumps up its processing tier to appear more "responsive".
- - If its in the middle of something, its processing tier gets locked to something between the two until they're done.
- Bots get that neato floating chat thing whenever they speak.
- Bots also play a sound when they speak too.
- All bots (except minebot, the mule, and Buddies) use a Secbot-style mover datum to scoot places.
- - So most bots now slide right to their destination instead of slowly inching over.
- - It also considers space as a valid movable tile. Probably just gonna keep this as a Floorbot exclusive.
- Most "stop whatever you're doing and give up" stuff has been moved to a single proc.
- Anything that a bot's supposed to do while moving has also been moved to a single proc.
- Most timed actions now use an actionbar.
- All bots have their ID access properly initialized. Currently gives butt and duckbots captain access, but that'll change.
- Most bots make a noise when they die.
- Changed most cooldown-y things to use ON_COOLDOWN.
- Most bot move speeds are vars, too.

Medibots:
 - They now use an actionbar to medicate
   - The increased processing rate makes them finish medicating people much, much faster.
   - The derelict shuttle bots are also much, much more dangerous, but also much more fair. And never seem to give up.
 - Hostile medbots'll now ignore targets if they have their poison(s) of choice in them.
   - This also means they won't waste their time pumping one person full of poison.
 - Emagged medbots have a chance on-emag to set their poison to a bunch of ???pill chems, or rarely a little bit of traitor poison.

Skullbots
 - They have the skeleton's voice. Clack.

Secbots
 - They now speak their lines, in addition to just blurting them out.
 - They shouldn't cheat anymore with their batons. One charge, one zap, stop breaking the law Beepsky!
 - Redid the secbot PDA program. Now everything's on one menu, and it'll send you messages confirming the thing you did!
 - They can be ordered to guard a specific area. They'll wander over there and then putter around until something stops them.
   - They can also be ordered to lockdown an area. Works the same as Guard, but they'll attack anyone in the area not wearing a head or sec ID. They wont cuff them, though.
   - Secbots can be set to not cuff people under a certain threat level, instead just stunning them and gloating.
 - Restored Beepsky's profanity-filled arrest-phrase, but with all the bad stuff bleeped out.
 - Secbots will instantly target anyone who attacks them, even if they're chasing someone else.
   - In addition, if there are nearby secbots, they'll also aggro onto that bot's attacker.
   - Beepsky will also send a PDA message to Sec that he's under attack.

Buttbots
 - They now contain the butt and arm they're built with.
 - They now scoot around aimlessly.
 - They now fart, using copypasted emote fartcode.
 - They release a little bit of oxygen when they fart
   - Or farts if they're emagged.
 - They remember whatever people say around them, then buttify random lines at their own pace.
 - Emagged buttbots will chase people down to fart on them.
   - They also don't stop farting and honestly I think I'm gonna tone it down a bit.
 - Emagged buttbots also have a chance to superfart.
   -They'll get plugged up, swell up with trapped wind, and then explode in a disgusting cloud of gas and gibs.
   - If they happen to do this on a bible, they go straight to hell.
 - They have a chance to gib on death, into either bloody or oily chunks.
 - They have some settings to disable most of their "behaviors".

Duckbots
 - They now mill around aimlessly
 - Emagged duckbots will chase people around.
 - They'll slowly "gestate" more eggs over time.
 - Occasionally, they'll decide to wander over to a random spot on the station.
 - Occasionally, *all* duckbots'll migrate to a random area on the station.
   - This can be controlled through packets. 
   - While moving to their destination, they get all access because I don't know how to restrict them to assistant-only areas.

Floorbots
 - They all use a shared list of in-use targets, claimed when seen and released when finished.
 - They set a home tile, and scan for valid tiles outward from there until they run out of valid tiles
   - Meaning they'll more reliably build (or destroy) tiles radiating out from a fixed point.
 - They no longer cause *massive awful lag* if there's a large group of them destroying the floor.
 - They should actually ignore inaccessible tiles more reliably.
 - They eat loose floor tiles and metal sheets by default, and should seek those things out if they're near its home.
   - These also happen instantly, as previously it just used a 0.5 second delay.
 - Their repair and disrepair procs use actionbars now. Makes it a lot more obvious they're actually doing something.
 - They're locked at a constant, mid-speed processing tier, as they're best designed to work in the background.

Cleanbots
 - Also uses a shared list of claimed targets.
   - They clear this every few minutes.
 - Uses a similar search method as Floorbots.
   - So they should keep to a specific area.
 - They now use an actionbar for cleaning. It also makes a sloshy noise!
 - They don't keep a list of tiles they've lubed. Instead, they just check if the floor is lubed and skip it if it is.
 
Firebots
 - More aggressively seeks out people to spray.
 - Increased the distance they'll try to spray at things to 3 tiles from 2.
 - They'll try to spray burning people who cross their proximity.
 - Their spray puddles now connect properly.
 - They now check a list of burning things instead of scanning every tile in view for things
   - Means they can see fire through walls, but honestly, its a firebot, looking for fire is kinda their thing

Buddies
 - Now use an actionbar to cuff
 - They now also announce their arrests to Research
 - They cuff a bit faster, too. Zipties, after all.
 - They continually shoot their toolgun at their target while they have one, regardless of processing.
   - Just about every tool has had their delay reduced significantly. Now they're actually really dangerous!
   - Flash modules flash every couple of seconds, and do the AoE flash if their target's out of range.
   - They also kinda spam their equipped gun. It'll ruin someone's day, but they tend to run out of power very quickly.

Chefbots
 - They now properly path to awful food things

Also dropped the ammofab's cost down to 1 TC. It's pretty frick, tbh.

## Why's this needed? <!-- Describe why you think this should be added to the game. -->

The bots are in need of some love, and I have more than enough to share with them. They generally suffer from feeling slow and marginally useful, even though their existing functionality is actually pretty robust when it actually works. That, and they're due for some cleanup so it's easier to love them in the future.

Also, making them do things more often and move around quicker gives the illusion of decent, functioning AI.

## Changelog
<!-- If necessary, put your changelog entry below. Otherwise, please delete it.
Use however you want to be credited in the changelog in place of CodeDude.
Use (*) for major changes and (+) for minor changes. For example: -->

```changelog
(u)Superlagg:
(*)Upgraded bot AI to be more responsive in general.
(*)Buddy combat AI has been upgraded to allow them to move *and* shoot.
(+)Secbots can be ordered to guard and lockdown specific areas.
(+)Floorbots now eat loose floor metal by default.
(+)Added actionbars to most times bot actions, like repairing floors and injecting people with drugs.
(+)Ammofab now costs 1 TC.
```

Superlagg commented:
Added a few by_categories and set firebots and secbots to use this instead of a view() check.

Secbots can be ordered to guard areas, either in person (just its current area), or via PDA (type in an area name).

Secbots can also be ordered to attack anyone who isnt sec or a head who is in that area.

Secbots are a lot more talkative. They speak what they're doing, for the most part, and telegraph their mode changes a bit more.

Secbots will send a PDA message when attacked (on a cooldown!), and will request help from nearby secbots.

Merged most secbot functions into a bunch of procs, cus otherwise it was fricked as heck.

Secbot control has gotten a facelift, now the controls are on the same screen as the bots, so you can order them around more conveniently.

Emagged duckbots will eventually get bored of annoying a certain person and find someone else to pester.

Added a proc to PDA programs that'll send a message to its owner's PDA.

Readded Beepsky's swear-filled arrest message, but with all the bad stuff bleeped out.

Superlagg commented:
Added more stuff:

Bots in general:
- They make jetpack particles if they move in space
  - yeah they're not supposed to be high-tech and fancy and whatnot, but I'd rather something be there for when they inevitably wander out into space than it just looking like they can walk on nothing.
- Replaced most view checks with by_type IN_RANGE checks.
  - So most bots can see through walls, but it speeds up performance.

Buttbots:
- Widened their emagged buttcabulary. Sort of.
- They have a chance to not forget the line they buttify.
 
Cleanbots:
- Clear the global claimed-by-cleanbot turf list every few minutes. Just so one bot doesn't claim it, get distracted, and forbid everyone else from cleaning it.
- They now actually claim and unclaim turfs properly.

Guardbots:
- No longer use speechbubbles due to them using complicated overlay stuff that I really don't want to get into right now.

Medbots:
- When emagged, they have a chance to, instead of panc, dose people with ???pill chems or, rarely, traitor poisons
- Added an override var list thing that'll force it to inject a specific chem via var-editing.

Secbots:
- Reduced chase distance to 13 tiles, but extended their pathing distance to reach just about anywhere in those 13 tiles.
- No longer runtime when talking.

Superlagg commented:
Updated stuff thanks to Tarm's feedback:
- Floorbots and cleanbots no longer refuse to rescan tiles on inner search-rings, so now they'll go back and do stuff to things that happen to show up near their search origin.
- Floorbots and cleanbots now store the coordinates of the turfs they claim and deem inaccessible. It now actually works.
  - Though large groups of cleanbots all suddenly becoming active on the same tile in a massive pile of filth will appear to all go after the same tile for a while. Dunno how to fix that, but after a while they'll spread out and start respecting each other's claims and space.
- Derelict medibots have been made more obnoxious.
  - They no longer get interrupted if they move. Instead, they just continue their action and chase after you.
  - If their action completes without anyone nearby, they just smoke whatever they were going to inject you with.
  - They now pick 1-4 chems from a list of medical-sounding garbage each injection attempt, including things like ecoli, mercury, atropine, and ketamine. Unlikely to kill you, but it'll definitely give you something to think about.
- Secbots actually follow orders to not cuff people now.

Adharainspace commented on _std/types.dm:
```diff
@@ -160,6 +160,9 @@ var/list/list/by_cat = list()
 #define TR_CAT_TELEPORT_JAMMERS "teleport_jammers"
+#define TR_CAT_BURNING_MOBS "dude_on_fire"
+#define TR_CAT_HOTSPOTS "air_on_fire"
+#define TR_CAT_BURNING_THINGS "stuff_on_fire"
```
```suggestion
#define TR_CAT_BURNING_ITEMS "items_on_fire"
```

Superlagg replied:
Done!

Adharainspace commented on _std/types.dm:
```diff
@@ -160,6 +160,9 @@ var/list/list/by_cat = list()
 #define TR_CAT_JOHNBILLS "johnbills"
 #define TR_CAT_OTHERBILLS "otherbills"
 #define TR_CAT_TELEPORT_JAMMERS "teleport_jammers"
+#define TR_CAT_BURNING_MOBS "dude_on_fire"
```
```suggestion
#define TR_CAT_BURNING_MOBS "dudes_on_fire"
```

Superlagg replied:
Done!

Adharainspace commented on _std/types.dm:
```diff
@@ -160,6 +160,9 @@ var/list/list/by_cat = list()
 #define TR_CAT_OTHERBILLS "otherbills"
 #define TR_CAT_TELEPORT_JAMMERS "teleport_jammers"
+#define TR_CAT_BURNING_MOBS "dude_on_fire"
+#define TR_CAT_HOTSPOTS "air_on_fire"
```
```suggestion
```
delete this, and use by_type for hotspots instead of by_cat.

Superlagg replied:
Did that, works just fine too!

Adharainspace commented on _std/pathfinding.dm:
```diff
@@ -215,6 +215,17 @@
+			if(!LinkBlockedWithAccess(src, T, ID))
+				L.Add(T)
+	return L
+
```
this proc looks good to me.

Superlagg replied:
=3

Tarmunora commented on code/modules/robotics/bot/medbot.dm:
```diff
@@ -38,7 +44,33 @@
+	var/override_reagent_limit_mult = 0.9
 	var/terrifying = 0 // for making the medbots all super fucked up
+	/// List of drugs that terrifying derelist bots will inject
+	var/static/list/terrifying_meds = list("formaldehyde" = 15,
+																				"ketamine" = 5,
+																				"pancuronium" = 5,
+																				"haloperidol" = 5,
+																				"atropine" = 30,
+																				"methamphetamine" = 20,
+																				"blood" = 20,
+																				"fog" = 20,
+																				"calomel" = 10,
+																				"ether" = 10,
+																				"mercury" = 5,
+																				"loose_screws" = 10,
+																				"salts1" = 10,
+																				"salmonella" = 10,
+																				"MRSA" = 10,
+																				"rat_venom" = 10,
+																				"green mucus" = 10,
+																				"mucus" = 10,
+																				"e.coli" = 10)
```
Not really a fan of a lot of these, especially the disease reagents.

Superlagg replied:
Okay I changed up most of these, replaced all the disease chems with more dangerous medical-sounding chems.

Tarmunora reviewed (changes requested):
Glancing through the guardbot changes, I think those might be better off cut entirely from this and potentially reproposed in a separate PR

Superlagg commented:
Turns out secbots were getting distracted by their own patrol script when going after targets, they'd chase someone for a few seconds, then just sort of give up and go back to patrol. Fixed that.

Then it turned out that if their patrol beacon had something dense on it, the bot wouldnt even try to move there. Now, every bot'll check for an open adjacent tile and head there instead.

Also, reverted the big changes with guardbots, I'll PR those things later since it made them a lot more threatening. I also gave them a gunfire cooldown, as their new processing rate made them spam it a bit.

While I was at it, I made the admin-only Assaultbuddy actually work. It now has the gun it's supposed to have, a bigger cell, and starts hostile to just about everything.

And I made the Buddycuff actionbar have the right owner. It was set to the task datum, which didn't make any sense. Now it works!

Superlagg commented:
Okay, I made some changes!
- Changed up the terrifying bot chemlist to remove the disease reagents, and replaced them with more harmful medical chems.
- Sped up the terrifying bot's injection delay from 2.5 seconds to 2.0 seconds.
- Made the bot able to ignore unreachable targets more reliably.
- Reduced the cooldown on the terrifying bot's same-target selection, so they'll more eagerly go after the same target after injecting them
- Emagged medbots have a 1% chance to reroll their poisons after injecting someone.

Tarmunora commented on code/modules/robotics/bot/medbot.dm:
```diff
@@ -38,7 +44,29 @@
+	var/override_reagent_limit_mult = 0.9
 	var/terrifying = 0 // for making the medbots all super fucked up
+	/// List of drugs that terrifying derelist bots will inject
+	var/static/list/terrifying_meds = list("formaldehyde" = 15,
+																				"ketamine" = 5,
+																				"pancuronium" = 5,
+																				"haloperidol" = 15,
+																				"morphine" = 20,
+																				"cold_medicine" = 40,
+																				"simethicone" = 10,
+																				"sulfonal" = 5, /* its an oldetimey sedative */
+																				"atropine" = 30,
+																				"methamphetamine" = 20,
+																				"ethanol" = 20, /* rubbing alcohol */
+																				"fog" = 20,
+																				"ether" = 10,
+																				"chlorine" = 10, /* disinfectant */
+																				"mercury" = 5)
```
```suggestion
	var/static/list/terrifying_meds = list("formaldehyde" = 15,
																				"ketamine" = 25,
																				"pancuronium" = 5,
																				"haloperidol" = 15,
																				"morphine" = 20,
																				"cold_medicine" = 40,
																				"simethicone" = 10,
																				"sulfonal" = 5, /* its an oldetimey sedative */
																				"atropine" = 10,
																				"methamphetamine" = 30,
																				"ethanol" = 20, /* rubbing alcohol */
																				"ether" = 10,
																				"chlorine" = 10, /* disinfectant */
																				"mercury" = 5)
```

Superlagg replied:
Sure! Though, the reason for all the atropine was that it was an overdose, but I suppose that it'd still be pretty debilitating like that!

Camryn-Buttes commented on code/modules/robotics/bot/medbot.dm:
```diff
@@ -474,110 +494,194 @@
+		src.path = null
+
+/datum/action/bar/icon/medbot_inject
+	duration = 30 SECONDS
```
This is far too long in practice.

Somewhere between 5 and 10 seconds seems like a better value

For patients with only one or two types of damage, this is *significantly* slower than the old behavior of making multiple injections spaced a couple seconds apart.

Luxizzle commented:
I'm wondering why the guard and lockdown options for secbots isnt a list instead of the current text input. Getting the name right or even remembering an area name can be a bit annoying at times, especially if you have to write it completely correct.

Superlagg commented:
> I'm wondering why the guard and lockdown options for secbots isnt a list instead of the current text input. Getting the name right or even remembering an area name can be a bit annoying at times, especially if you have to write it completely correct.

Mainly because the list would include every area on the station, which tends to be very, very long. I think I included a function on the program to send you a list of station areas... though i might've taken out out because it seemed like a bad idea for some reason. I'll have to find a way to make it easier to use!

Thanks for the feedback!

Luxizzle commented:
If there was, i didn't see it.

Also another thing i noticed, the secbots seem to get stuck on things like artifacts while patrolling. I think this was already an issue that happened before though, but i just noticed it.

Overall i really like the changes and I'm glad the bots got some love!
